src/charm.py

```diff
@@ -150,8 +149,9 @@ class GrafanaAgentServiceError(GrafanaAgentError):
 
 
 @trace_charm(
-    tracing_endpoint="tracing_endpoint",
-    server_cert="server_cert_path",
+    # these attrs are implemented on GrafanaAgentCharm
+    tracing_endpoint="_charm_tracing_endpoint",
+    server_cert="_server_cert",
     extra_types=(COSAgentRequirer, GrafanaAgentCharm, JujuTopology, SnapFstab),
```

Comment:
I'm not 100% sure we need to add `GrafanaAgentCharm` here, given it's a superclass. IIRC `trace_charm` will discover attributes on the superclasses as well without needing to redecorate them. In fact, I'm afraid this will result in duplicate nested spans since we're effectively decorating it twice.

Reply:
Just checked, I'm right.